### PR TITLE
[TT-7482] feat: implement graph sql pump

### PR DESCRIPTION
**Pull request by kofoworola**
## Description

This PR adds the Graph SQL pump, a pump that parses graphql records and saves it to SQL storage that will be used by the user and the dashboard for displaying analytics. Link to the story [here](https://tyktech.atlassian.net/browse/TT-7482) for more details

## Related Issue

[TT-7482](https://tyktech.atlassian.net/browse/TT-7482)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

[TT-7482]: https://tyktech.atlassian.net/browse/TT-7482?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by buraksezer — APPROVED** [review]

File: pumps/graph_sql.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+	g.log.Debug("pump initialized and table set up")
+	return nil
+}
+
+func (g *GraphSQLPump) WriteData(ctx context.Context, data []interface{}) error {
+	g.log.Debug("Attempting to write ", len(data), " records...")
+
+	var graphRecords []*analytics.GraphRecord
+	for _, r := range data {
```
Comment (tbuchaillot) [inline pumps/graph_sql.go:94]: @kofoworola what do you think on moving this conversion logic to a separated function to reduce cognitive complexity? Something like:
```
func (g *GraphSQLPump) getGraphRecords(data []interface{}) ([]*analytics.GraphRecord, error) {
    var graphRecords []*analytics.GraphRecord
    for _, r := range data {
        if r != nil {
            rec, ok := r.(analytics.AnalyticsRecord)
            if !ok || !rec.IsGraphRecord() {
                continue
            }
            gr, err := rec.ToGraphRecord()
            if err != nil {
                g.log.Warnf("error converting 1 record")
                g.log.WithError(err).Debug("error converting record")
                continue
            }
            graphRecords = append(graphRecords, &gr)
        }
    }
    return graphRecords, nil
}
```

File: pumps/graph_sql.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func (g *GraphSQLPump) New() Pump {
+	return &GraphSQLPump{}
+}
+
+func (g *GraphSQLPump) Init(conf interface{}) error {
+	g.log = log.WithField("prefix", GraphSQLPrefix)
+
+	if err := mapstructure.Decode(conf, &g.SQLConf); err != nil {
```
Comment (tbuchaillot) [inline pumps/graph_sql.go:37]: @kofoworola we are missing a way to set the things thru env variables. Please check how SQL pump does it
https://github.com/TykTechnologies/tyk-pump/blob/master/pumps/sql.go#L136
https://github.com/TykTechnologies/tyk-pump/blob/master/pumps/sql.go#L105

File: pumps/graph_sql_test.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	"time"
+
+	"github.com/TykTechnologies/tyk-pump/analytics"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGraphSQLPump_Init(t *testing.T) {
```
Comment (tbuchaillot) [inline pumps/graph_sql_test.go:17]: Can we add a scenario to set the basic conf thru env vars?

File: pumps/graph_sql_test.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+		rec := baseRecord
+		rec.TimeStamp = timestamp
+		rec.Month = timestamp.Month()
+		rec.Day = timestamp.Day()
+		rec.Year = timestamp.Year()
+		records = append(records, rec)
+		expectedTables = append(expectedTables, fmt.Sprintf("%s_%s", conf.TableName, timestamp.Format("20060102")))
+	}
+
+	r.NoError(pump.WriteData(context.Background(), records))
```
Comment (tbuchaillot) [inline pumps/graph_sql_test.go:281]: should we clean up the table after this?

**Review by tbuchaillot — CHANGES_REQUESTED** [review]
@kofoworola GREAT job man! left some small comments. Also, it would be amazing if you can add some piece with example in the README.

File: pumps/graph_sql_test.go (hunk 4 above)
Comment (kofoworola) [reply pumps/graph_sql_test.go:313]: Added cleanup in L305

File: pumps/graph_sql_test.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+			r.Equalf(len(tc.responses), len(resultRecords), "responses count do no match")
+			if diff := cmp.Diff(expectedResponses, resultRecords, cmpopts.IgnoreFields(analytics.GraphRecord{}, "AnalyticsRecord")); diff != "" {
+				t.Error(diff)
+			}
+		})
+	}
+}
+
+func TestGraphSQLPump_Sharded(t *testing.T) {
+	r := require.New(t)
```
[the remark this answers is not in the document]
Comment (kofoworola) [reply pumps/graph_sql_test.go:268]: assert doesn't end a test if the condition doesn't pass, which is needed in some cases, and that is what require does

File: pumps/sql.go [hunk 6]
```diff
@@ -70,6 +70,10 @@ type SQLConf struct {
 	// Specifies the amount of records that are going to be written each batch. Type int. By
 	// default, it writes 1000 records max per batch.
 	BatchSize int `json:"batch_size" mapstructure:"batch_size"`
+	// TableName is a configuration field unique to the sql-graph pump, this field specifies
+	// the name of the sql table to be created/used for the pump in the cases of non-sharding
+	// in the case of sharding, it specifies the table prefix
+	TableName string `json:"table_name" mapstructure:"table_name"`
```
[the remark this answers is not in the document]
Comment (kofoworola) [reply pumps/sql.go:76]: I was thinking it might be needed in the SQL conf at some point, but it is better to just extend it.

**Review by tbuchaillot — APPROVED** [review]
LGTM

File: analytics/graph_record.go [hunk 7]
```diff
@@ -94,33 +94,35 @@ func (a *AnalyticsRecord) ToGraphRecord() (GraphRecord, error) {
...
 			return record, err
 		}
-		record.HasErrors = true
+		errBytes, t, _, err := jsonparser.Get(dat, "errors")
+		if err != nil && err != jsonparser.KeyPathNotFoundError {
+			log.WithError(err).Error("error getting response errors")
+			return record, err
+		}
+		if t != jsonparser.NotExist {
+			if err := json.Unmarshal(errBytes, &record.Errors); err != nil {
```
[the remark this answers is not in the document]
Comment (kofoworola) [reply analytics/graph_record.go:120]: yup those blocks can be combined, thanks!

[the remark this answers is not in the document]
Comment (kofoworola) [reply analytics/graph_record.go:120]: Yeah for verbosity, went with explicitly checking both errors and type

[the remark this answers is not in the document]
Comment (kofoworola) [reply analytics/graph_record.go:123]: I added a few comment to better explain that block.

**Pull request merged**